### ISSUE TITLE
Does Jupyterhub live reload the config file if it changes on disk?

### ISSUE THREAD

**Issue by tlvu** [jupyterhub/jupyterhub#2541]
I am using Jupyterhub with dockerspawner.  Jupyterhub runs in its own docker container (image [jupyterhub/jupyterhub](https://hub.docker.com/r/jupyterhub/jupyterhub)) and each user logged-in will spawn a new docker container for their chosen notebook image (ex: public [jupyter/scipy-notebook](https://hub.docker.com/r/jupyter/scipy-notebook) image or one of our in-house notebook image with our specific packages installed).

In `/srv/jupyterhub/jupyterhub_config.py` I have something like this to give the user the selection of notebook images that can be used:
```
c.DockerSpawner.image_whitelist = {'notebook1': 'notebook1:1.0.0,                                                                     
                                   'notebook2': 'notebook2:1.1.1',
                                   'jupyter/scipy-notebook': 'jupyter/scipy-notebook:latest',
                                   }
```
If I release a new version of `notebook1`, I will need to update `/srv/jupyterhub/jupyterhub_config.py` with the new version of `notebook1`.  Will Jupyterhub notice the config file `/srv/jupyterhub/jupyterhub_config.py` has changed and reload it live or I have to restart Jupyterhub for the change to take effect?  Is there a way to trigger a manual live config reload?

If I restart Jupyterhub and at the same time I have existing users logged-in and doing work, will they lose their non-saved work?  Will their authentication become invalidated and they have to login again?

In general, how does Jupyterhub recommend deploying config change that is least disruptive to a massive number of actively logged-in users?

**Comment by minrk** [comment]
> Is there a way to trigger a manual live config reload?

No, the Hub must be restarted to reload configuration.

> If I restart Jupyterhub and at the same time I have existing users logged-in and doing work, will they lose their non-saved work?

This will depend on configuration, but usually no. This will depend on the values for `JupyterHub.cleanup_servers` and `Proxy.should_start`. By default, these are both True, in which case shutting down the Hub will shut down everything. However, in a production deployment, these should typically both be False, so that restarting the Hub has no disruption for users who are already running their servers at all.

> Will their authentication become invalidated and they have to login again?

Generally not, no.

> In general, how does JupyterHub recommend deploying config change that is least disruptive to a massive number of actively logged-in users?

Deploying a config change is always restarting the Hub process. By deploying an externally managed proxy (e.g. via systemd/supervisor/docker/kubernetes) and setting `c.JupyterHub.cleanup_servers = False`, there should be minimal disruption while the Hub restarts.

**Comment by tlvu** [comment]
@minrk Thanks for the answers.

Just to be sure, I run Jupyterhub in a docker container.  Jupyterhub is not physically installed on a machine like a traditional deployment.  Furthermore, I spawn a new docker container for a notebook server environment for each user.  Are all your answers and configurations (`JupyterHub.cleanup_servers`, `Proxy.should_start`, `c.JupyterHub.cleanup_servers = False`) applicable to my case?

Running things in container instead of directly on a machine make configuration harder, especially state persistence between container destruction and respawn (destroying the Jupyterhub container and respawn it).  Are there extra configurations I need to set to make this happen?

**Comment by minrk** [comment]
Yeah, for minimal disruption, best practice is:

1. put proxy in separate container from Hub (requires Proxy.should_start = False and setting CONFIGPROXY_AUTH_TOKEN env in both hub and proxy containers)
2. set c.JupyterHub.cleanup_servers = False (leaves users running while Hub restarts)

**Comment by tlvu** [comment]
@minrk I ended up not needing `Proxy.should_start = False` and `c.JupyterHub.cleanup_servers = False` since the notebooks servers are running in separate containers than in the same container as JupyterHub.

Once the JupyterHub container is destroyed and restarted, the user can simply resume working as if nothing happened.

While the JupyterHub container was down, users was not able to work but their session are kept intact so when the JupyterHub came back, user can continue working as if JupyterHub was never down.

I had to persist JupyterHub data for this resume to work (`c.JupyterHub.cookie_secret_file = '/persist/jupyterhub_cookie_secret'` and `c.JupyterHub.db_url = '/persist/jupyterhub.sqlite'` where `/persist` is a volume mount so it does not get destroyed when the JupyterHub container is destroyed).

So I am satisfied with this behavior since the bleep where JupyterHub is gone is not too long.  Closing this issue unless you have something even less disruptive for the users to let me know.  Thanks.
